server/datastore/mysql/mysql.go

```diff
@@ -214,11 +216,31 @@ func New(config config.MysqlConfig, c clock.Clock, opts ...DBOption) (*Datastore
 		clock:             c,
 		config:            config,
 		readReplicaConfig: options.replicaConfig,
+		writeCh:           make(chan itemToWrite),
 	}
 
+	go ds.writeChanLoop()
+
 	return ds, nil
 }
 
+type itemToWrite struct {
+	ctx   context.Context
+	errCh chan error
+	item  interface{}
+}
+
+func (d *Datastore) writeChanLoop() {
+	for item := range d.writeCh {
+		switch actualItem := item.item.(type) {
+		case *fleet.Host:
+			item.errCh <- d.SaveHost(item.ctx, actualItem)
```

Comment:
If the `errCh` is closed in the `defer` of the Serial functions, this will panic (write to a closed channel). This could happen if the send to `writeCh` succeeds before `ctx.Done()`, but `ctx.Done()` triggers before the send to `errCh`.

Reply:
Right, so the approach would be to close it after the write to the channel:

```
		case *fleet.Host:
			item.errCh <- d.SaveHost(item.ctx, actualItem)
			close(item.errCh)
```

?

Reply:
I'd make it buffered (`make(chan error, 1)`) so that it doesn't block on send (as the receiver might be gone by that time due to the context expiring), and wouldn't bother closing it. Another option would be to send it in a `select` with a default case if it is closed, but really we only ever send one value over it, so having it buffered would fix that.

Reply:
Makes sense, thank you!